[SPARK-18819][CORE] Double byte alignment on ARM platforms

michaelkamprath opened the pull request:
## What changes were proposed in this pull request?

On 32-bit ARM platforms, doubles require 8-byte alignment. The use of Java's `sun.misc.Unsafe` library for memory block management exposes Spark to misaligned doubles when using `org.apache.spark.unsafe.Platform.getDouble()` to extract a double from a memory block with arbitrary offset locations. For the alignment sensitive platforms, this proposed change creates an properly aligned buffer that the bytes to be cast to a double are first copied to, then the double the double is extracted via the `sun.misc.Unsafe.getDouble()` method. On all other platforms the previous behavior is maintained, which was to cast the double value from the memory block regardless of offset alignment . 

## How was this patch tested?

All unit tests pass, plus tested on both x86 and ARM71 platforms. For the ARM71 platform, it is verified that this fix removes the problem as described in SPARK-18819.

michaelkamprath committed:
[SPARK-18819] Double alignment on ARM

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -44,6 +49,9 @@
 
   private static final boolean unaligned;
+
+  private static final MemoryBlock _doubleBuffer;
```
This is a globally shared variable -- that's not going to work is it? reads would all corrupt each other.

michaelkamprath replied:
Admittedly, I am not deeply familiar with how this class is being using concurrently within a single JVM. If multiple threads are using it concurrently, you are right and a different approach must be found. I was trying to avoid the cost of reallocating a buffer with ever call to `getDouble()`.

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -119,7 +127,12 @@ public static void putFloat(Object object, long offset, float value) {
 
   public static double getDouble(Object object, long offset) {
-    return _UNSAFE.getDouble(object, offset);
+    if ( null == _doubleBuffer) {
```
Style nit:

```
if (_doubleBuffer == null) {
```

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -244,6 +257,34 @@ public static void throwException(Throwable t) {
+        byte[] heapObj = new byte[16];
+        long offset = BYTE_ARRAY_OFFSET;
+        long bufferSize = 16;
+        for (long i = 0; i < 8; ++i ) {
```
Spacing:

```
for (long i = 0; i < 8; i++) {
  if ((offset + i) % 8 == 0) {
```

Why a long?

michaelkamprath replied:
Simply to match the `offset` variable, which needs to be a long.

srowen reviewed (changes requested):
I think the question will be whether this can be supported with virtually 0 overhead to the x64 case.

kiszk commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -119,7 +127,12 @@ public static void putFloat(Object object, long offset, float value) {
+      return _UNSAFE.getDouble(object, offset);
+    } else {
+      copyMemory(object, offset, _doubleBuffer.getBaseObject(), _doubleBuffer.getBaseOffset(), 8);
+      return _UNSAFE.getDouble(_doubleBuffer.getBaseObject(), _doubleBuffer.getBaseOffset());
```
If `getLong()` never causes an alignment error, how about Double.longBitsToDouble(_UNSAFE.getLong()).  
While I am not familiar with ARM architecture, which one is faster?

michaelkamprath replied:
I will test this idea and report back. Thanks!

kiszk commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -244,6 +257,34 @@ public static void throwException(Throwable t) {
+            offset += i;
+            bufferSize -= i;
+            break;
+          }
```
Can we assign `null` into _doubleBuffer if `offset % 8 == 0`?

michaelkamprath replied:
No, we can't, because we would still need the buffer to avoid doing the direct double read from  an unaligned memory location on [line 131](https://github.com/michaelkamprath/spark/blob/30c6c997ca8e864a31d2f4dfa55d47b5aa629596/common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java#L131). 

The reason for finding an aligned offset here is because on ARM7, the actual memory address for the `byte[0]` location is not the memory address of the `byte[]` object. That starting point for the 0-index item is indicated by `Platform.BYTE_ARRAY_OFFSET`, which is the Java overhead memory used for managing the `byte[]` object. So if we read from `byte[0]`, that could be unaligned because of the Java overhead, hence the reason to find the index in the byte buffer where overhead+index is aligned. I am depending on the Java behavior of aligning on objects here, the `byte[]` object) for needing to only check that the offset is also aligned.

kiszk commented:
While we ideally expect `static final` is evaluated as a constant by a JIT compiler, I do not know the fact. I would appreciate someone would look at the binary code generated by OpenJDK since I cannot do it due to some reasons.

michaelkamprath committed:
[SPARK-18819] address pull request comments

michaelkamprath commented:
I've refactored this issue's solution using the idea that @kiszk supplied above. This ended up being a better idea, thanks. It produced cleaner code than my original allocated buffer approach. Also, based on the input that this solution should have zero impact on the x86 platform, I've refactored to a functor design pattern which creates the equivalent performance behavior for x86 as before this fix. 

The performance impact of the using an intermediate long as a buffer when fetching a double from a memory block looks to be about 10%. I ran some performance tests writing and reading random doubles to random  aligned locations in a buffer 100M times over. On my mac laptop and an ARM unit, the performance difference was the long-buffered approach was about 10% slower than the original direct fetch approach. However, since the this refactor isolates the fetching behavior into a functor, the 10% slowdown only affects the ARM platform.

I also added some unit tests if only to convince myself I hadn't broken anything.

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/DoubleAccessBuffered.java:
```diff
@@ -0,0 +1,33 @@
+
+import sun.misc.Unsafe;
+
+public class DoubleAccessBuffered implements DoubleAccessFunctor {
```
(Why public?)

michaelkamprath replied:
No reason but forgot to uncheck the Eclipse modifiers when creating a new class. Fixed.

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/DoubleAccessDirect.java:
```diff
@@ -0,0 +1,33 @@
+
+import sun.misc.Unsafe;
+
+class DoubleAccessDirect implements DoubleAccessFunctor {
```
Implementations should be `final` for completeness and to maximize chance of JIT inlining.

michaelkamprath replied:
fixed

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -244,6 +251,19 @@ public static void throwException(Throwable t) {
+        _doubleAccessFunc = new DoubleAccessBuffered();
+      } else {
+        // set the buffer address to null to indicate no buffer allocated
+        _doubleAccessFunc = new DoubleAccessDirect();;
```
extra semicolon

michaelkamprath replied:
fixed

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -44,6 +48,9 @@
 
   private static final boolean unaligned;
+
+  private static final DoubleAccessFunctor _doubleAccessFunc;
```
Java/Scala wouldn't prefix members with _

michaelkamprath replied:
changed

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -244,6 +251,19 @@ public static void throwException(Throwable t) {
+            arch);
+        _doubleAccessFunc = new DoubleAccessBuffered();
+      } else {
+        // set the buffer address to null to indicate no buffer allocated
```
The comment seems wrong

michaelkamprath replied:
removed (vestige from prior approach)

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -252,6 +272,7 @@ public static void throwException(Throwable t) {
       LONG_ARRAY_OFFSET = 0;
       FLOAT_ARRAY_OFFSET = 0;
       DOUBLE_ARRAY_OFFSET = 0;
+      _doubleAccessFunc = null;
```
Is this supposed to be null now rather than just the other implementation?

michaelkamprath replied:
I am setting it to `null` to keep in line with the behavior of the `_UNSAFE` class variable. In case, `_UNSAFE` is null. In that case, any calls to the `Platform.get...` or `Platform.put...` will result in a `NullPointerException`. I'm assuming it is a programming error elsewhere to call these methods without checking ( [example](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/memory/MemoryManager.scala#L200) ).

srowen commented on common/unsafe/src/test/java/org/apache/spark/unsafe/PlatformUtilSuite.java:
```diff
@@ -74,4 +76,71 @@ public void memoryDebugFillEnabledInTest() {
+    // write a double to an unaligned location
+    long unalignedOffset = testBuffer.getBaseOffset() + 3;
+    // double check unalignment just to be sure:
+    if (unalignedOffset%8 == 0) {
```
Space around operators like %

michaelkamprath replied:
fixed

srowen commented on common/unsafe/src/test/java/org/apache/spark/unsafe/PlatformUtilSuite.java:
```diff
@@ -74,4 +76,71 @@ public void memoryDebugFillEnabledInTest() {
+    long unalignedOffset = testBuffer.getBaseOffset() + 3;
+    // double check unalignment just to be sure:
+    if (unalignedOffset%8 == 0) {
+      ++unalignedOffset;
```
x++ not ++x unless there is a specific need for preincrement

michaelkamprath replied:
fixed

srowen commented on common/unsafe/src/test/java/org/apache/spark/unsafe/PlatformUtilSuite.java:
```diff
@@ -74,4 +76,71 @@ public void memoryDebugFillEnabledInTest() {
+          buffered.getDouble(unsafe, testBuffer.getBaseObject(), alignedOffset),
+          0.000001);
+    } catch (Throwable cause) {
+      // can't run test because Unsafe is unavailable.
```
You can't do this in a test, it will catch any error. Construct the test to deal with this case explicitly.

michaelkamprath replied:
refactored test to isolate the `Unsafe` object creation better.

srowen reviewed (changes requested):
Still needs a fair bit of work, but might be more feasible. I am not sure about the overhead of the extra dispatch here unless we can get it to inline reliably. The hit for x64 may be too much.

kiszk commented on common/unsafe/src/main/java/org/apache/spark/unsafe/DoubleAccessFunctor.java:
```diff
@@ -0,0 +1,25 @@
+
+import sun.misc.Unsafe;
+
+interface DoubleAccessFunctor {
```
Is there any reason to use `interface`? I think that `abstract class` is enough for this.  
I like virtual call rather than interface call. This is because [the code sequnce](https://wiki.openjdk.java.net/display/HotSpot/VirtualCalls) for virtual call is simpler than [that](https://wiki.openjdk.java.net/display/HotSpot/InterfaceCalls) for interface call.

michaelkamprath replied:
No reason other than there was no commonality between the concrete classes other than the interface. I will switch to an abstract class given your performance input.

srowen replied:
Interesting @kiszk I had not encountered this before. I would have used an interface for simplicity, but I understand this is an argument about performance in hot code. Design-wise it's virtually identical. According to http://stackoverflow.com/questions/21608240/performance-difference-between-passing-interface-and-class-reloaded this particular case might be just as fast. I think we can leave it unless someone has a strong feeling otherwise.

michaelkamprath committed:
[SPARK-18819] address more pull request comments

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -22,10 +22,14 @@
 import sun.misc.Cleaner;
 import sun.misc.Unsafe;
 
 public final class Platform {
```
BTW there is a whole second copy of this class under the 'sketch' module that I think would also have to be modified. Hm, then the new classes need to be copied too, ugh

michaelkamprath replied:
I missed that. I'll address that when we determine the final path here (per my comment below).

srowen commented on common/unsafe/src/test/java/org/apache/spark/unsafe/PlatformUtilSuite.java:
```diff
@@ -74,4 +76,74 @@ public void memoryDebugFillEnabledInTest() {
+      unsafeField.setAccessible(true);
+      unsafe = (sun.misc.Unsafe) unsafeField.get(null);
+    } catch (Throwable cause) {
+      unsafe = null;
```
Could just return here to skip the test?

srowen reviewed (commented):
I'm still sort of uneasy about the complexity and tiny overhead in this path. Dumb question, what's the use case for running Spark on ARM? I still think of that as a mobile architecture.

kiszk commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -244,6 +251,18 @@ public static void throwException(Throwable t) {
+
+      // determine whether double access should be aligned.
+      String arch = System.getProperty("os.arch", "");
+      if (arch.matches("^(arm|arm32)")) {
```
What's happen on ARM 64-bit?

michaelkamprath replied:
@kiszk I have tested on ARM 64 (`aarch64`). [Any alignment works for double access there](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.den0024a/ch08s02.html), though 8-byte aligned access looks to be about 10% faster than unaligned access. Using an intermediate long buffer (your idea) is about 5% slower than direct access regardless of being aligned or not. In both cases, I tested with an ODROID C2 using Oracle Java 8.

kiszk replied:
Thanks for your clarification. I was afraid that ARM 64 may return `arm`.

michaelkamprath commented:
@srowen To answer the use case question, it is primarily academic for learning and testing. Students and researchers build clusters of Raspberry PI, ODROID, or other SBCs to have a cost effective access to a multi-node hardware cluster. [Here](http://likemagicappears.com/projects/raspberry-pi-cluster/) [are](http://coen.boisestate.edu/ece/research-areas/raspberry-pi/) [some](https://www.raspberrypi.org/magpi/pi-spark-supercomputer/) [examples](http://hackaday.com/2016/05/09/designing-a-high-performance-parallel-personal-cluster/) [of](http://katie.atomicburn.com/2016/06/12/2016-school-gt-exhibition-raspberry-piodroid-c2-supercomputer/) [projects](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4803722/). There is even [a commercial vendor](https://www.picocluster.com/collections/) selling these SBC clusters. [In my own case](http://diybigdata.net/odroid-xu4-cluster/), its being used to economically learn how to deal problems of efficiency (it's easier to spot and work through patterns of inefficiency on constrained systems than full powered systems). 

I am personally not aware if there are currently any server-class CPUs that requires double alignment. Double alignment SPARC processors used to be the bane of my existence in the early 2000's, but that was over a decade ago. My understanding is that today x86 supports unaligned double access with [a theoretical performance hit](https://developers.redhat.com/blog/2016/06/01/how-to-avoid-wasting-megabytes-of-memory-a-few-bytes-at-a-time/) that [in practice is rarely seen](http://lemire.me/blog/2012/05/31/data-alignment-for-speed-myth-or-reality/). Typically you never concern yourself with alignment in Java because the JVM takes care of it for you, but here we are delving into the world of Unsafe, which bypasses the protections the JVM provides. Admittedly, it took me a long while to even figure out that my problem was related to alignment because as indicated, I haven't dealt with such issues in over a decade.

With all that said, maybe a better approach here is to create a patch that users can use to create a spark build when they want to run Spark on a system that requires double alignment, which to the best of my knowledge are currently just ARM 32-bit CPUs. That would even let to be more concise, without needing to determine at runtime which method to use. And if ever should arise a server-class CPU with alignment requirements, we know what to do.

Given that

srowen commented:
I can't imagine Raspberry Pi is a good way to run Spark. It won't be cost effective compared to a simple commodity server. Yeah, I'm concerned that this introduces overhead for every other architecture in a hot path, and I am not sure this actually makes Spark work on such an architetcure anyway.

michaelkamprath commented:
@srowen No one has any delusions that an ARM platform is the place to do heavy lifting for data. But if your goal is academic in nature (hands on learning how to set up and operate a cluster,  learning the API, etc), it's a very economical platform. I do understand your concerns for the core target and totally get why you wouldn't want this fix in the main line, so I will instead pursue a patch approach for those who'd be interested. I'm not sure what that means here ... do you or I close this pull request?

michaelkamprath committed:
[SPARK-18819] refactor as standalone patch

Refactoring this fix as a standalone patch that can be used to
create custom build of Spark that run on ARM32 platforms. Since
these changes will not be merged into the main line, removed
conditional usage of aligned or unaligned double access and made
this code only us aligned double access.

michaelkamprath commented:
I am withdrawing this pull request and will instead publish a patch for those who are interested in using Spark on a platforms requiring byte alignment for double access.
